client/pooling.zig

```diff
@@ -0,0 +1,179 @@
+const std = @import("std");
+
+const log = std.log.scoped(.pooling);
+
+pub fn SimpleKey(comptime Child: type) type {
+    return struct {
+        value: Child,
+
+        const Self = @This();
+
+        pub fn compare(self: Self, other: Self) std.math.Order {
+            return if (self.value == other.value) .eq else .lt;
+        }
+    };
+}
+
+pub fn SizedKey(comptime Child: type) type {
+    return struct {
+        value: Child,
+        size: usize,
+
+        const Self = @This();
+
+        pub fn compare(self: Self, other: Self) std.math.Order {
+            if (self.size > other.size) {
+                return .lt; // less than means it won't fit
+            }
+
+            if (self.size == other.size) {
+                return .eq; // equals means it's an exact match
+            }
+
+            if (self.size < other.size) {
+                return .gt; // greater than means it's a match, but not a perfect match
+            }
+
+            unreachable;
+        }
+    };
+}
+
+pub const StringKey = struct {
+    value: []const u8,
+
+    pub fn compare(self: StringKey, other: StringKey) std.math.Order {
+        return if (std.mem.eql(u8, self.value, other.value)) .eq else .lt;
+    }
+};
+
+pub fn FrameReferencedResourcePool(comptime Context: type, comptime Key: type, comptime Value: type, comptime create_val: anytype, comptime release_val: anytype, comptime frames_to_keep_entry: comptime_int) type {
+    // errors to make this very clear on what to do
+    comptime {
+        if (std.meta.activeTag(@typeInfo(Key)) != .@"struct")
+            @compileError("Key type must be a struct. There are examples of key types in this source file.");
+
+        if (!@hasDecl(Key, "compare"))
+            @compileError("Key type is missing an 'eql' function. There are examples of key types in this source file.");
+
+        if (!@hasField(Key, "value"))
+            @compileError("Key type is missing the 'value' field. There are examples of key types in this source file.");
```

Comment:
I don't believe these are needed anymore.

```suggestion
        if (!@hasDecl(Key, "compare"))
            @compileError("Key type is missing a 'compare' function. There are examples of key types in this source file.");
```